src/TokenHandler/CreditRegistry.mo

```diff
@@ -0,0 +1,101 @@
+import RBTree "mo:base/RBTree";
+import Principal "mo:base/Principal";
+import Time "mo:base/Time";
+import Option "mo:base/Option";
+import Iter "mo:base/Iter";
+
+import Journal "Journal";
+
+module {
+  public type StableData = (Int, [(Principal, Int)]);
+
+  /// Tracks credited funds (usable balance) associated with each principal.
+  public class CreditRegistry(
+    journal : Journal.Journal,
+    isFrozen : () -> Bool,
+  ) {
+    var map : RBTree.RBTree<Principal, Int> = RBTree.RBTree<Principal, Int>(Principal.compare);
+
+    /// Total sum of credited funds in the credit registry.
+    var creditTotal_ : Int = 0;
+
+    /// Retrieves the total credited funds in the credit registry.
+    public func creditTotal() : Int = creditTotal_;
+
+    /// Gets the current credit amount associated with a specific principal.
+    public func get(p : Principal) : Int = Option.get(map.get(p), 0);
+
+    /// Deducts amount from P’s usable balance.
+    /// The flag `strict` enables checking the availability of sufficient funds.
+    func debit_(p : Principal, amount : Nat, strict : Bool) : Bool {
+      if (isFrozen()) {
+        return false;
+      };
+
+      let ?currentCredit = map.get(p) else return false;
```

Comment:
`map.get(p) == null` means that the credit is 0. `return false` is only correct if `strict` is true. If `strict` is false then we should put a negative value into the map.

Reply:
Fixed